irksome/base_time_stepper.py

```diff
@@ -0,0 +1,155 @@
+import numpy
+
+from abc import abstractmethod
+from firedrake import Function, NonlinearVariationalProblem, NonlinearVariationalSolver
+from firedrake.dmhooks import pop_parent, push_parent
+from firedrake.petsc import PETSc
+from .tools import AI, get_stage_space, getNullspace
+from pyop2.types import MixedDat
+
+
+class BaseTimeStepper:
+    def __init__(self, F, t, dt, u0,
+                 bcs=None, appctx=None, nullspace=None):
+        self.F = F
+        self.t = t
+        self.dt = dt
+        self.u0 = u0
+        if bcs is None:
+            bcs = ()
+        self.orig_bcs = bcs
+        self.nullspace = nullspace
+        self.V = u0.function_space()
+
+        appctx_base = {
+            "F": F,
+            "t": t,
+            "dt": dt,
+            "u0": u0,
+            "bcs": bcs,
+            "nullspace": nullspace}
+
+        if appctx is None:
+            self.appctx = appctx_base
+        else:
+            self.appctx = {**appctx, **appctx_base}
+
+    @abstractmethod
+    def advance(self):
+        pass
+
+    @abstractmethod
+    def solver_stats(self):
+        pass
+
+
+# Stage derivative + stage value + (maybe?) RadauIIAIMEX
+class StageCoupledTimeStepper(BaseTimeStepper):
+    def __init__(self, F, t, dt, u0, num_stages,
+                 bcs=None, solver_parameters=None,
+                 appctx=None, nullspace=None,
+                 splitting=None, bc_type="DAE",
+                 butcher_tableau=None, bounds=None):
+
+        super().__init__(F, t, dt, u0,
+                         bcs=bcs, appctx=appctx, nullspace=nullspace)
+
+        self.num_stages = num_stages
+        if butcher_tableau:
+            assert num_stages == butcher_tableau.num_stages
+            self.appctx["butcher_tableau"] = butcher_tableau
+        if splitting is None:
+            splitting = AI
+        self.splitting = splitting
+        self.appctx["splitting"] = splitting
+        self.bc_type = bc_type
+
+        self.num_steps = 0
+        self.num_nonlinear_iterations = 0
+        self.num_linear_iterations = 0
```

Comment:
If this were to cover `RadauIIAIMEX`, we'd need more iteration counters here, to deal with Iterators vs. Propagators.

Reply:
They could probably be handled directly in the imex class?

Reply:
But we're not going to merge this now.